File: scripts/train_model.py
#!/usr/bin/env python3
"""
Script de entrenamiento para el modelo CNN de Pokémon
Requiere: pip install tensorflow numpy pillow
"""

import tensorflow as tf
import numpy as np
import json
import os

# Configuración
IMG_SIZE = 128
NUM_CLASSES = 15
BATCH_SIZE = 32
EPOCHS = 10

MODEL_DIR = os.path.join(os.path.dirname(__file__), '..', 'public', 'pokemon-model')
os.makedirs(MODEL_DIR, exist_ok=True)

# Lista de Pokémon
POKEMON_LIST = [
    {"id": 1, "name": "Bulbasaur"},
    {"id": 2, "name": "Ivysaur"},
    {"id": 3, "name": "Venusaur"},
    {"id": 4, "name": "Charmander"},
    {"id": 5, "name": "Charmeleon"},
    {"id": 6, "name": "Charizard"},
    {"id": 7, "name": "Squirtle"},
    {"id": 8, "name": "Wartortle"},
    {"id": 9, "name": "Blastoise"},
    {"id": 25, "name": "Pikachu"},
    {"id": 26, "name": "Raichu"},
    {"id": 133, "name": "Eevee"},
    {"id": 143, "name": "Snorlax"},
    {"id": 150, "name": "Mewtwo"},
    {"id": 151, "name": "Mew"}
]

def create_light_model():
    """Crea un modelo CNN ligero para clasificación de Pokémon"""
    model = tf.keras.Sequential([
        # Capa de entrada
        tf.keras.layers.Input(shape=(IMG_SIZE, IMG_SIZE, 3)),

        # Bloque convolucional 1
        tf.keras.layers.Conv2D(8, (3, 3), activation='relu'),
        tf.keras.layers.MaxPooling2D((2, 2)),

        # Bloque convolucional 2
        tf.keras.layers.Conv2D(16, (3, 3), activation='relu'),
        tf.keras.layers.MaxPooling2D((2, 2)),

        # Bloque convolucional 3
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu'),
        tf.keras.layers.MaxPooling2D((2, 2)),

        # Aplanar
        tf.keras.layers.Flatten(),

        # Capa densa
        tf.keras.layers.Dense(64, activation='relu'),
        tf.keras.layers.Dropout(0.3),

        # Capa de salida
        tf.keras.layers.Dense(NUM_CLASSES, activation='softmax')
    ])

    return model

def generate_mock_data(samples_per_class=30):
    """Genera datos sintéticos de entrenamiento"""
    print(f"Generando {samples_per_class} muestras por clase...")

    X = []
    y = []

    for class_idx in range(NUM_CLASSES):
        for i in range(samples_per_class):
            # Crear imagen sintética con color característico de cada clase
            base_color = np.random.normal(class_idx * 0.1, 0.2, (IMG_SIZE, IMG_SIZE, 3))
            noise = np.random.normal(0, 0.1, (IMG_SIZE, IMG_SIZE, 3))
            image = np.clip(base_color + noise, 0, 1)

            X.append(image)
            y.append(class_idx)

    X = np.array(X, dtype=np.float32)
    y = tf.keras.utils.to_categorical(y, NUM_CLASSES)

    return X, y

def train_model():
    """Entrena el modelo y lo guarda"""
    print("=" * 50)
    print("🚀 ENTRENAMIENTO MODELO POKÉDEX CNN")
    print("=" * 50)

    # Crear modelo
    print("\n📐 Creando modelo...")
    model = create_light_model()

    model.compile(
        optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
        loss='categorical_crossentropy',
        metrics=['accuracy']
    )

    model.summary()

    # Generar datos
    print(f"\n📊 Generando datos de entrenamiento...")
    X_train, y_train = generate_mock_data(30)
    print(f"   Total de muestras: {len(X_train)}")

    # Entrenar
    print(f"\n🏋️  Entrenando por {EPOCHS} épocas...")
    history = model.fit(
        X_train, y_train,
        epochs=EPOCHS,
        batch_size=BATCH_SIZE,
        validation_split=0.2,
        verbose=1
    )

    # Mostrar resultados
    final_acc = history.history['accuracy'][-1]
    final_val_acc = history.history['val_accuracy'][-1]
    print(f"\n📈 Resultados:")
    print(f"   Precisión entrenamiento: {final_acc:.1%}")
    print(f"   Precisión validación: {final_val_acc:.1%}")

    # Guardar en formato TensorFlow.js
    print(f"\n💾 Guardando modelo...")

    # Guardar con tfjs converter format
    model.save(os.path.join(MODEL_DIR, 'model.keras'))

    # Convertir a formato TF.js
    print("   Convirtiendo a formato TensorFlow.js...")

    # Usar tensorflowjs_converter
    try:
        import tensorflowjs as tfjs
        tfjs.converters.save_keras_model(model, MODEL_DIR)
        print("   ✅ Conversión exitosa")
    except ImportError:
        print("   ⚠️  tensorflowjs no instalado, guardando solo formato Keras")
        print("   Para convertir: pip install tensorflowjs")
        print(f"   Luego: tensorflowjs_converter --input_format=keras {MODEL_DIR}/model.keras {MODEL_DIR}")

    # Crear metadata
    metadata = {
        "version": "1.0.0",
        "modelName": "Pokedex CNN",
        "numClasses": NUM_CLASSES,
        "imageSize": IMG_SIZE,
        "classes": POKEMON_LIST,
        "accuracy": {
            "train": float(final_acc),
            "validation": float(final_val_acc)
        },
        "createdAt": tf.timestamp().numpy().tolist()
    }

    with open(os.path.join(MODEL_DIR, 'metadata.json'), 'w') as f:
        json.dump(metadata, f, indent=2)

    print(f"\n✅ Modelo guardado en: {MODEL_DIR}")

    # Listar archivos
    for file in os.listdir(MODEL_DIR):
        size = os.path.getsize(os.path.join(MODEL_DIR, file))
        print(f"   📄 {file} ({size/1024:.1f} KB)")

    print("\n" + "=" * 50)
    print("✅ ENTRENAMIENTO COMPLETADO")
    print("=" * 50)

if __name__ == '__main__':
    train_model()

File: scripts/train-model.js
import * as tf from '@tensorflow/tfjs'
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

// Configurar backend CPU
await tf.setBackend('cpu')
await tf.ready()

const MODEL_DIR = path.join(__dirname, '..', 'public', 'pokemon-model')
const IMG_SIZE = 224

// Lista de Pokémon (16 iniciales)
const POKEMON_LIST = [
  { id: 1, name: 'Bulbasaur' },
  { id: 2, name: 'Ivysaur' },
  { id: 3, name: 'Venusaur' },
  { id: 4, name: 'Charmander' },
  { id: 5, name: 'Charmeleon' },
  { id: 6, name: 'Charizard' },
  { id: 7, name: 'Squirtle' },
  { id: 8, name: 'Wartortle' },
  { id: 9, name: 'Blastoise' },
  { id: 25, name: 'Pikachu' },
  { id: 26, name: 'Raichu' },
  { id: 133, name: 'Eevee' },
  { id: 143, name: 'Snorlax' },
  { id: 150, name: 'Mewtwo' },
  { id: 151, name: 'Mew' }
]

const NUM_CLASSES = POKEMON_LIST.length

// Crear directorio si no existe
if (!fs.existsSync(MODEL_DIR)) {
  fs.mkdirSync(MODEL_DIR, { recursive: true })
}

// Crear modelo CNN
function createModel() {
  const model = tf.sequential()

  // Capa convolucional 1
  model.add(tf.layers.conv2d({
    inputShape: [IMG_SIZE, IMG_SIZE, 3],
    filters: 32,
    kernelSize: 3,
    activation: 'relu'
  }))
  model.add(tf.layers.maxPooling2d({ poolSize: 2 }))
  model.add(tf.layers.batchNormalization())

  // Capa convolucional 2
  model.add(tf.layers.conv2d({
    filters: 64,
    kernelSize: 3,
    activation: 'relu'
  }))
  model.add(tf.layers.maxPooling2d({ poolSize: 2 }))
  model.add(tf.layers.batchNormalization())

  // Capa convolucional 3
  model.add(tf.layers.conv2d({
    filters: 128,
    kernelSize: 3,
    activation: 'relu'
  }))
  model.add(tf.layers.maxPooling2d({ poolSize: 2 }))
  model.add(tf.layers.batchNormalization())

  // Capa convolucional 4
  model.add(tf.layers.conv2d({
    filters: 256,
    kernelSize: 3,
    activation: 'relu'
  }))
  model.add(tf.layers.maxPooling2d({ poolSize: 2 }))
  model.add(tf.layers.batchNormalization())

  // Aplanar
  model.add(tf.layers.flatten())

  // Capa densa 1
  model.add(tf.layers.dense({
    units: 512,
    activation: 'relu'
  }))
  model.add(tf.layers.dropout({ rate: 0.5 }))

  // Capa densa 2
  model.add(tf.layers.dense({
    units: 256,
    activation: 'relu'
  }))
  model.add(tf.layers.dropout({ rate: 0.3 }))

  // Capa de salida
  model.add(tf.layers.dense({
    units: NUM_CLASSES,
    activation: 'softmax'
  }))

  return model
}

// Generar datos sintéticos para entrenamiento (simulación)
function generateMockData(numSamples = 100) {
  const xs = []
  const ys = []

  for (let i = 0; i < numSamples; i++) {
    // Generar imagen aleatoria
    const image = tf.randomNormal([IMG_SIZE, IMG_SIZE, 3])
    xs.push(image)

    // Etiqueta aleatoria
    const label = Math.floor(Math.random() * NUM_CLASSES)
    const oneHot = new Array(NUM_CLASSES).fill(0)
    oneHot[label] = 1
    ys.push(oneHot)
  }

  return {
    xs: tf.stack(xs),
    ys: tf.tensor2d(ys)
  }
}

// Entrenar modelo
async function trainModel() {
  console.log('🚀 Iniciando entrenamiento del modelo Pokémon...')
  console.log(`📊 Clases: ${NUM_CLASSES} Pokémon`)
  console.log(`🖼️  Tamaño de imagen: ${IMG_SIZE}x${IMG_SIZE}`)

  const model = createModel()

  model.compile({
    optimizer: tf.train.adam(0.0001),
    loss: 'categoricalCrossentropy',
    metrics: ['accuracy']
  })

  console.log('\n📐 Arquitectura del modelo:')
  model.summary()

  // Generar datos de entrenamiento sintéticos
  console.log('\n📥 Generando datos de entrenamiento...')
  const { xs, ys } = generateMockData(500)

  // Entrenar
  console.log('\n🏋️  Entrenando...')
  await model.fit(xs, ys, {
    epochs: 10,
    batchSize: 16,
    validationSplit: 0.2,
    callbacks: {
      onEpochEnd: (epoch, logs) => {
        console.log(`Época ${epoch + 1}/10 - Pérdida: ${logs.loss.toFixed(4)} - Precisión: ${(logs.acc * 100).toFixed(2)}%`)
      }
    }
  })

  // Guardar modelo
  console.log('\n💾 Guardando modelo...')
  await model.save(`file://${MODEL_DIR}`)

  // Crear archivo de metadatos
  const metadata = {
    version: '1.0.0',
    numClasses: NUM_CLASSES,
    imageSize: IMG_SIZE,
    pokemon: pokemonData.map(p => ({
      id: p.id,
      name: p.name,
      index: pokemonData.indexOf(p)
    })),
    createdAt: new Date().toISOString()
  }

  fs.writeFileSync(
    path.join(MODEL_DIR, 'metadata.json'),
    JSON.stringify(metadata, null, 2)
  )

  console.log('\n✅ Modelo entrenado y guardado exitosamente!')
  console.log(`📁 Ubicación: ${MODEL_DIR}`)
  console.log(`📄 Archivos: model.json, weights.bin, metadata.json`)

  // Liberar memoria
  xs.dispose()
  ys.dispose()
  model.dispose()
}

trainModel().catch(err => {
  console.error('❌ Error entrenando:', err)
  process.exit(1)
})
